### PR TITLE
Move diagnostic service to LSP layer

### PR DESCRIPTION
**Pull request by tmat**
Move diagnostic service and all code that depends on it out of Features layer to LSP layer. 
This code is currently client code - it does not run OOP but also does not have dependency on editor. The code will be further simplified/eliminated once we transition to pull diagnostics.

Split AnalyzerHelpers helpers among AnalyzerOptionsExtensions in Features, AnalyzerOptionsExtensions in Analyzers shared project, DiagnosticAnalyzerExtensions and DocumentAnalysisExecutor.

Removes IIncrementalAnalyzer2.

Simplify CodeFixService.FixAllDiagnosticProvider.

Move ClassificationOptionsStorage, CompletionOptionStorage and SignatureHelpOptionsStorage to LSP layer.

Fixes https://github.com/dotnet/roslyn/issues/59184
Fixes https://github.com/dotnet/roslyn/issues/45936

Requires:
- [x] https://github.com/dotnet/roslyn/pull/58765
- [x] https://github.com/dotnet/roslyn/pull/59316
- [x] https://github.com/dotnet/roslyn/pull/59751
- [x] https://devdiv.visualstudio.com/DevDiv/_git/TypeScript-VS/pullrequest/382162

### CODE REVIEW

File: src/Analyzers/Core/Analyzers/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs [hunk 1]
```diff
@@ -83,9 +83,6 @@ public override void Initialize(AnalysisContext context)
             });
         }
 
-        [PerformanceSensitive(
-            "https://github.com/dotnet/roslyn/issues/23583",
-            Constraint = nameof(AnalyzerHelper.GetOption) + " is expensive and should be avoided if a syntax-based fast path exists.")]
```
[the remark this answers is not in the document]
Comment (tmat) [reply src/Analyzers/Core/Analyzers/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:88]: Reverted.

**Review by dibarbet — APPROVED** [review]
mostly looks OK to me, would recommend Manish looking as well.

**Comment by dibarbet** [comment]
This should most likely have a val build as well

File: src/EditorFeatures/Test/Diagnostics/DiagnosticAnalyzerServiceTests.cs [hunk 2]
```diff
@@ -1246,7 +1246,7 @@ private static async Task RunAllAnalysisAsync(IIncrementalAnalyzer analyzer, Tex
                 await analyzer.AnalyzeSyntaxAsync(document, InvocationReasons.Empty, CancellationToken.None).ConfigureAwait(false);
                 await analyzer.AnalyzeDocumentAsync(document, bodyOpt: null, reasons: InvocationReasons.Empty, cancellationToken: CancellationToken.None).ConfigureAwait(false);
             }
-            else if (analyzer is IIncrementalAnalyzer2 analyzer2)
+            else if (analyzer is IIncrementalAnalyzer analyzer2)
```
Comment (mavasani) [inline src/EditorFeatures/Test/Diagnostics/DiagnosticAnalyzerServiceTests.cs:1249]:
```suggestion
            else
```

File: src/EditorFeatures/Test/Diagnostics/DiagnosticAnalyzerServiceTests.cs [hunk 3]
```diff
@@ -1246,7 +1246,7 @@ private static async Task RunAllAnalysisAsync(IIncrementalAnalyzer analyzer, Tex
...
             {
                 await analyzer2.AnalyzeNonSourceDocumentAsync(textDocument, InvocationReasons.Empty, CancellationToken.None).ConfigureAwait(false);
```
Comment (mavasani) [inline src/EditorFeatures/Test/Diagnostics/DiagnosticAnalyzerServiceTests.cs:1251]:
```suggestion
                await analyzer.AnalyzeNonSourceDocumentAsync(textDocument, InvocationReasons.Empty, CancellationToken.None).ConfigureAwait(false);
```

File: src/Features/Core/Portable/SolutionCrawler/AggregateIncrementalAnalyzer.cs [hunk 4]
```diff
@@ -136,7 +136,7 @@ public async Task RemoveProjectAsync(ProjectId projectId, CancellationToken canc
         public async Task NonSourceDocumentOpenAsync(TextDocument textDocument, CancellationToken cancellationToken)
         {
             if (TryGetAnalyzer(textDocument.Project, out var analyzer) &&
-                analyzer is IIncrementalAnalyzer2 analyzer2)
+                analyzer is IIncrementalAnalyzer analyzer2)
```
Comment (mavasani) [inline src/Features/Core/Portable/SolutionCrawler/AggregateIncrementalAnalyzer.cs:139]: Redundant check

File: src/Features/Core/Portable/SolutionCrawler/AggregateIncrementalAnalyzer.cs [hunk 5]
```diff
@@ -145,7 +145,7 @@ public async Task NonSourceDocumentOpenAsync(TextDocument textDocument, Cancella
         public async Task NonSourceDocumentCloseAsync(TextDocument textDocument, CancellationToken cancellationToken)
         {
             if (TryGetAnalyzer(textDocument.Project, out var analyzer) &&
-                analyzer is IIncrementalAnalyzer2 analyzer2)
+                analyzer is IIncrementalAnalyzer analyzer2)
```
Comment (mavasani) [inline src/Features/Core/Portable/SolutionCrawler/AggregateIncrementalAnalyzer.cs:148]: Redundant check (also at all places below).

File: src/Features/Core/Portable/SolutionCrawler/IncrementalAnalyzerBase.cs [hunk 6]
```diff
@@ -46,5 +46,26 @@ public virtual Task RemoveDocumentAsync(DocumentId documentId, CancellationToken
...
+
+        public virtual Task AnalyzeNonSourceDocumentAsync(TextDocument textDocument, InvocationReasons reasons, CancellationToken cancellationToken)
+            => Task.CompletedTask;
+
+        public void LogAnalyzerCountSummary()
+        {
+        }
+
+        /// <summary>
+        /// Order all incrementl analyzers below DiagnosticIncrementalAnalyzer
```
Comment (mavasani) [inline src/Features/Core/Portable/SolutionCrawler/IncrementalAnalyzerBase.cs:67]:
```suggestion
        /// Order all incremental analyzers below DiagnosticIncrementalAnalyzer
```

File: src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.NormalPriorityProcessor.cs [hunk 7]
```diff
@@ -485,7 +485,7 @@ static async Task DocumentResetAsync(IIncrementalAnalyzer analyzer, TextDocument
                             {
                                 await analyzer.DocumentResetAsync(document, cancellationToken).ConfigureAwait(false);
                             }
-                            else if (analyzer is IIncrementalAnalyzer2 analyzer2)
+                            else if (analyzer is IIncrementalAnalyzer analyzer2)
                             {
                                 await analyzer2.NonSourceDocumentResetAsync(textDocument, cancellationToken).ConfigureAwait(false);
```
Comment (mavasani) [inline src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.NormalPriorityProcessor.cs:490]:
```suggestion
                            else
                            {
                                await analyzer.NonSourceDocumentResetAsync(textDocument, cancellationToken).ConfigureAwait(false);
```

File: src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.NormalPriorityProcessor.cs [hunk 8]
```diff
@@ -497,7 +497,7 @@ static async Task AnalyzeSyntaxAsync(IIncrementalAnalyzer analyzer, TextDocument
                             {
                                 await analyzer.AnalyzeSyntaxAsync(document, reasons, cancellationToken).ConfigureAwait(false);
                             }
-                            else if (analyzer is IIncrementalAnalyzer2 analyzer2)
+                            else if (analyzer is IIncrementalAnalyzer analyzer2)
                             {
                                 await analyzer2.AnalyzeNonSourceDocumentAsync(textDocument, reasons, cancellationToken).ConfigureAwait(false);
```
Comment (mavasani) [inline src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.NormalPriorityProcessor.cs:502]:
```suggestion
                            else
                            {
                                await analyzer.AnalyzeNonSourceDocumentAsync(textDocument, reasons, cancellationToken).ConfigureAwait(false);
```

File: src/Features/LanguageServer/Protocol/Features/CodeFixes/CodeFixService.FixAllDiagnosticProvider.cs [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+            private readonly bool _includeSuppressedDiagnostics;
+
+            public FixAllDiagnosticProvider(IDiagnosticAnalyzerService diagnosticService, ImmutableHashSet<string> diagnosticIds)
+            {
+                _diagnosticService = diagnosticService;
+
+                // When computing FixAll for unnecessary pragma suppression diagnostic,
+                // we need to include suppressed diagnostics, as well as reported compiler and analyzer diagnostics.
+                // A null value for 'diagnosticIdsForDiagnosticProvider' passed to 'FixAllDiagnosticProvider'
+                // ensures the latter.
```
Comment (mavasani) [inline src/Features/LanguageServer/Protocol/Features/CodeFixes/CodeFixService.FixAllDiagnosticProvider.cs:31]:
```suggestion
                // A null value for '_diagnosticIds' ensures the latter.
```

File: src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.cs [hunk 10]
```diff
@@ -222,6 +222,8 @@ private static DiagnosticAnalysisResult GetResultOrEmpty(ImmutableDictionary<Dia
         public void LogAnalyzerCountSummary()
             => _telemetry.ReportAndClear(_correlationId);
 
+        public int Priority => 0;
```
Comment (mavasani) [inline src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.cs:225]: Probably add a comment that `DiagnosticIncrementalAnalyzer` has the highest priority (lowest int value) amongst all incremental analyzers (others have priority 1).

File: src/Workspaces/Core/Portable/SolutionCrawler/IIncrementalAnalyzer2.cs [hunk 11]
```diff
@@ -1,26 +0,0 @@
...
-#nullable disable
-
-using System.Threading;
-using System.Threading.Tasks;
-
-namespace Microsoft.CodeAnalysis.SolutionCrawler
-{
-    // CONSIDER: We can merge IIncrementalAnalyzer2 with IIncrementalAnalyzer once all of our
-    //           IVT partners that use IIncrementalAnalyzer have migrated to ExternalAccess layer.
-    internal interface IIncrementalAnalyzer2 : IIncrementalAnalyzer
```
Comment (mavasani) [inline src/Workspaces/Core/Portable/SolutionCrawler/IIncrementalAnalyzer2.cs:14]: This PR should also close #45936. I'll edit the PR description.

**Pull request merged**